protocol/binary/stream_reader.go

```diff
@@ -0,0 +1,445 @@
+	if _, err := sr.read(bs); err != nil {
+		return false, err
+	}
+
+	switch bs[0] {
+	case 0:
+		return false, nil
+	case 1:
+		return true, nil
+	default:
+		return false, decodeErrorf("invalid bool value: %q", bs[0])
+	}
+}
+
+// ReadInt8 reads a Thrift encoded int8 value.
+func (sr *StreamReader) ReadInt8() (int8, error) {
+	bs := sr.buffer[0:1]
+	_, err := sr.read(bs)
+	return int8(bs[0]), err
+}
+
+// ReadInt16 reads a Thrift encoded int16 value.
+func (sr *StreamReader) ReadInt16() (int16, error) {
+	bs := sr.buffer[0:2]
+	_, err := sr.read(bs)
+	return int16(bigEndian.Uint16(bs)), err
+}
+
+// ReadInt32 reads a Thrift encoded int32 value.
+func (sr *StreamReader) ReadInt32() (int32, error) {
+	bs := sr.buffer[0:4]
+	_, err := sr.read(bs)
+	return int32(bigEndian.Uint32(bs)), err
+}
+
+// ReadInt64 reads a Thrift encoded int64 value.
+func (sr *StreamReader) ReadInt64() (int64, error) {
+	bs := sr.buffer[0:8]
+	_, err := sr.read(bs)
+	return int64(bigEndian.Uint64(bs)), err
+}
+
+// ReadString reads a Thrift encoded string.
+func (sr *StreamReader) ReadString() (string, error) {
+	bs, err := sr.ReadBinary()
+	return string(bs), err
+}
+
+// ReadDouble reads a Thrift encoded double, returning a float64.
+func (sr *StreamReader) ReadDouble() (float64, error) {
+	val, err := sr.ReadInt64()
+	return math.Float64frombits(uint64(val)), err
+}
+
+// ReadBinary reads a Thrift encoded binary type, returning a byte array.
+func (sr *StreamReader) ReadBinary() ([]byte, error) {
+	length, err := sr.ReadInt32()
+	if err != nil {
+		return nil, err
+	}
+
+	if length < 0 {
+		return nil, decodeErrorf("negative length %v specified for binary field", length)
+	}
+
+	if length == 0 {
+		return []byte{}, nil
+	}
+
+	if length > bytesAllocThreshold {
+		var buf bytes.Buffer
+		_, err := io.CopyN(&buf, sr.reader, int64(length))
+		if err == io.EOF {
+			// All EOFs are unexpected when streaming
+			err = io.ErrUnexpectedEOF
+		}
+
+		return buf.Bytes(), err
+	}
+
+	bs := make([]byte, length)
+	_, err = sr.read(bs)
+	return bs, err
+}
+
+// ReadStructBegin reads the "beginning" of a Thrift encoded struct.  Since
+// there is no encoding for the beginning of a struct, this is a noop.
+func (sr *StreamReader) ReadStructBegin() error {
+	return nil
+}
+
+// ReadStructEnd reads the "end" of a Thrift encoded struct.  Since
+// `ReadFieldBegin` will already be interpreting field-type of whether it's a
+// stop field or not, there is no real representation of a struct's end, making
+// this a noop.
+func (sr *StreamReader) ReadStructEnd() error {
+	return nil
+}
+
+// ReadFieldBegin reads off a Thrift encoded field-header returning that and a
+// 'bool' representing whether or not a field-value follows.
+// A 'false' without any error means that it has reached the stop-field.  There
+// is no guarantee that the field-header is valid in this case.
+func (sr *StreamReader) ReadFieldBegin() (stream.FieldHeader, bool, error) {
+	fh := stream.FieldHeader{}
+
```

Comment:
optional: given how we're using `fh`, named returns might be more shorter, and also document the bool:
```suggestion
func (sr *StreamReader) ReadFieldBegin() (fh stream.FieldHeader, ok bool, err error) {
```